dev/scripts/release.py

```diff
@@ -0,0 +1,121 @@
+    'composer.lock',
+    'CONTRIBUTING.md',
+    'docker-compose.yaml',
+    'Dockerfile.phpdoc',
+    'nginx.example',
+    'package.json',
+    'package-lock.json',
+    'phpstan.neon',
+    'postinstall.js',
+    'README.md',
+    'release.sh',
+    'yarn-postinstall.js',
+    'yarn.lock',
+    'SECURITY.md',
+    'semantic.json',
+    'web.config.example',
+]
+TARGET_DIR = 'release'
+
+
+def create_archives(source_dir: str, archive_name: str):
+    zip_command = [
+        'zip',
+        '-r',  # Recursive
+        '-q',  # Quiet
+        f'{TARGET_DIR}/nameless-{archive_name}.zip',
+        '.',
+        *itertools.chain.from_iterable((('-x', f'{name}/*') for name in EXCLUDE_DIRS)),
+        *itertools.chain.from_iterable((('-x', f'{name}') for name in EXCLUDE_FILES)),
+    ]
+
+    tar_command = [
+        'tar',
+        '-c',  # Compress
+        '-J',  # Use xzip
+        '--owner=0',  # Set owner inside archive to root
+        '--group=0',
+        '-f', f'{TARGET_DIR}/nameless-{archive_name}.tar.xz',
+        *[f'--exclude={name}' for name in [*EXCLUDE_DIRS, *EXCLUDE_FILES]],
+        '.',
+    ]
+
+    print(f'Creating zip archive: {archive_name}')
+    subprocess.check_call(zip_command,
+                          shell=False)
+
+    print(f'Creating tar.xz archive: {archive_name}')
+    subprocess.check_call(tar_command,
+                          env={'XZ_DEFAULTS': '-T 0'},  # Enable multithreading
+                          shell=False)
+
+
+if __name__ == '__main__':
+    if not Path('.git').exists():
+        print('.git does not exists')
+        sys.exit(1)
+
+    print('Deleting vendor files')
+    shutil.rmtree('core/assets/vendor', ignore_errors=True)
+    shutil.rmtree('node_modules', ignore_errors=True)
+    shutil.rmtree('vendor', ignore_errors=True)
+
+    # Create base archive
+    create_archives('.', 'base')
+
+    # Run npm and composer (production dependencies only)
+    subprocess.check_call(['npm', 'ci', '-q', '--cache', '.node_cache'],
+                          stdout=PIPE)
+    subprocess.check_call(['composer', 'install', '--no-dev', '--no-interaction'],
+                          stdout=PIPE)
+    create_archives('.', 'deps-dist')
+
+    # Run composer again, to install development dependencies
+    subprocess.check_call(['composer', 'install', '--no-interaction'],
+                          stdout=PIPE)
+    create_archives('.', 'deps-dev')
+
+    # Create archive with files changed since last update
+
+    upgrade_temp = Path('release', 'upgrade_temp')
+
+    # Find previous tag
+    previous_tag_command = ['git', 'describe', '--abbrev=0', '--tags']
+    previous_tag = subprocess.check_output(previous_tag_command, shell=False)[:-1].decode()
+
+    # Find all files changed between previous tag and HEAD (current state)
+    changed_command = ['git', 'diff', previous_tag, 'HEAD', '--name-only', '--diff-filter=d']
+    changed_files = subprocess.check_output(changed_command, shell=False)[:-1].decode()
```

Comment:
When I tested this locally it seemed to take all of the files... for this to work properly does it have to be run on the `develop` branch right after a new tag (ie `2.0.3`) was made?

Reply:
The upgrade package should include all files changed since the previous tag (so the tag for the upcoming release should not be created yet). I tested it on the `release/2.0.3` branch and it generated all changed files since the `2.0.2` tag.